source/particle/particle_handler.cc

```diff
@@ -222,6 +261,661 @@ namespace aspect
+                                                       std::multimap<types::LevelInd,Particle <dim> >     &received_particles,
+                                                       const std::vector<std::vector<active_cell_it> >    &send_cells)
+    {
+      // Determine the communication pattern
+      const std::vector<types::subdomain_id> neighbors (triangulation->ghost_owners().begin(),
+                                                        triangulation->ghost_owners().end());
+      const unsigned int n_neighbors = neighbors.size();
+
+      Assert(n_neighbors == send_particles.size(),
+             ExcMessage("The particles to send to other processes should be sorted into a vector "
+                        "containing as many vectors of particles as there are neighbor processes. This "
+                        "is not the case for an unknown reason. Contact the developers if you encounter "
+                        "this error."));
+
+      unsigned int n_send_particles = 0;
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        n_send_particles += send_particles[i].size();
+
+      if (n_send_particles == 0)
+        return;
+
+      const unsigned int cellid_size = sizeof(CellId::binary_type);
+      const unsigned int particle_size = begin()->size() + size_callback() + cellid_size;
+
+      // Determine the amount of data we will send to other processors
+      std::vector<unsigned int> n_send_data(n_neighbors);
+      std::vector<unsigned int> n_recv_data(n_neighbors);
+
+      std::vector<unsigned int> send_offsets(n_neighbors);
+      std::vector<unsigned int> recv_offsets(n_neighbors);
+
+      // Allocate space for sending and receiving particle data
+      std::vector<char> send_data(n_send_particles * particle_size);
+      void *data = static_cast<void *> (&send_data.front());
+
+      for (types::subdomain_id neighbor_id = 0; neighbor_id < n_neighbors; ++neighbor_id)
+        {
+          send_offsets[neighbor_id] = reinterpret_cast<std::size_t> (data) - reinterpret_cast<std::size_t> (&send_data.front());
+
+          // Copy the particle data into the send array
+          typename std::vector<particle_iterator>::const_iterator
+          particle = send_particles[neighbor_id].begin(),
+          end_particle = send_particles[neighbor_id].end();
+          for (unsigned int i=0; i<send_particles[neighbor_id].size(); ++i)
+            {
+              // If no target cells are given, use the iterator information
+              active_cell_it cell;
+              if (send_cells.size() == 0)
+                cell = send_particles[neighbor_id][i]->get_surrounding_cell(*triangulation);
+              else
+                cell = send_cells[neighbor_id][i];
+
+              const CellId::binary_type cellid = cell->id().template to_binary<dim>();
+              memcpy(data, &cellid, cellid_size);
+              data = static_cast<char *>(data) + cellid_size;
```

Comment:
it would be nicer to use the boost serialization facilities here, but that can wait for another day.